[JENKINS-25623] time out & kill infinite recursion

kohsuke opened the pull request:
[JENKINS-25623](https://issues.jenkins-ci.org/browse/JENKINS-25623)

kohsuke commented:
Depends on https://github.com/cloudbees/groovy-cps/pull/39
@reviewbybees

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

jglick commented:
Does this extend pause (#32) to non-`Step` code? If so, there is no test coverage for it.

jglick commented on src/main/java/org/jenkinsci/plugins/workflow/cps/Safepoint.java:
```diff
@@ -0,0 +1,30 @@
+ *
+ * @author Kohsuke Kawaguchi
+ */
+// classes compiled from Pipeline Script refers to this class, so it needs to be public, but
```
…but what?

jglick replied:
`@Restricted(NoExternalUse.class)`?

kohsuke replied:
yes

jglick commented on src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThread.java:
```diff
@@ -151,62 +152,59 @@ public StepExecution getStep() {
     @Nonnull Outcome runNextChunk() throws IOException {
         assert program!=null;
 
-        while (true) {
```
Reviewers need to look [here](https://github.com/jenkinsci/workflow-cps-plugin/pull/46/files?w=1#diff-1a70e438dd28d7a6307fc2bfb55904e8L154). Basically almost all of this was just reindentation.

jglick commented on src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java:
```diff
@@ -221,96 +219,52 @@ public void unexport(BodyReference ref) {
-        final Future<Future<?>> f;
+        final SettableFuture<Void> f = SettableFuture.create();
         try {
-        f = runner.submit(new Callable<Future<?>>() {
```
[reindentation](https://github.com/jenkinsci/workflow-cps-plugin/pull/46/files?w=1#diff-6452cd1bea46404d75cc1ba4559cc3adL226)

jglick commented on src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java:
```diff
@@ -347,56 +301,66 @@ public boolean isPaused() {
+    private boolean run() throws IOException {
+        boolean changed = false;
         boolean ending = false;
-        do {
```
[reindentation](https://github.com/jenkinsci/workflow-cps-plugin/pull/46/files?w=1#diff-6452cd1bea46404d75cc1ba4559cc3adL357)

jglick commented on src/test/java/org/jenkinsci/plugins/workflow/WorkflowJobNonRestartingTest.java:
```diff
@@ -236,4 +238,25 @@ public void addAction() throws Exception {
         assertNotNull(p.getAction(WhoAmI.class));
     }
+
+    @Test
```
`@Issue`

jglick commented on src/test/java/org/jenkinsci/plugins/workflow/WorkflowJobNonRestartingTest.java:
```diff
@@ -236,4 +238,25 @@ public void addAction() throws Exception {
+        b.doStop(); // abort, abort!
+
+        f.get(3, TimeUnit.SECONDS); // wait for the result to come in
+        assertEquals(Result.ABORTED, b.getResult());
```
Suffices to write

```java
jenkins.assertBuildStatus(Result.ABORTED, jenkins.waitForCompletion(b));
```

jglick commented on src/test/java/org/jenkinsci/plugins/workflow/WorkflowJobNonRestartingTest.java:
```diff
@@ -236,4 +238,25 @@ public void addAction() throws Exception {
+        QueueTaskFuture<WorkflowRun> f = p.scheduleBuild2(0);
+        WorkflowRun b = f.getStartCondition().get(3, TimeUnit.SECONDS);
+
+        going.block(3000);    // get the buld going, which will loop infinitely
```
There is a race condition here, that the test thread might continue after `going` is called but before the loop is entered. Whether that matters or not, I am not sure.

kohsuke replied:
FTR, this has since been addressed by change in the definition of the script.

jglick commented:
:ant: the JIRA summary claims that this change makes `timeout` work on an infinite loop (the original motivating use case), yet there is no test proving that this actually happens—only that an _external_ abort works.

Other than that, :bee: though I scarcely understand the implementation.

kohsuke commented:
Addressed the reviewer comment. Assuming the :bee: still stands, I'll leave it up to @jglick to decide the right moment to merge this, so that I won't rock any boat.

jglick commented:
re:bee:

jglick commented on src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java:
```diff
@@ -347,56 +301,66 @@ public boolean isPaused() {
 
-        if (doneSomeWork) {
+        if (changed) {
             saveProgram();
```
Caused some regressions in programs which were legitimately broken. I am more concerned that the constant saving may have caused some performance regressions. Perhaps we need to throttle saves.
